pilot/pkg/config/coredatamodel/incrementalController.go

```diff
@@ -0,0 +1,374 @@
+		if oldEpVersion != newEpVersion {
+			if err := c.edsUpdate(conf); err != nil {
+				log.Warnf("edsUpdate: %v", err)
+			}
+		}
+	}
+
+	c.configStoreMu.Lock()
+	c.configStore = configs
+	c.configStoreMu.Unlock()
+
+	if svcChanged {
+		c.options.XDSUpdater.ConfigUpdate(true)
+	}
+}
+
+func (c *IncrementalController) incrementalUpdate(resources []*sink.Object) {
+	svcChanged := c.serviceChange(resources)
+	for _, obj := range resources {
+		conf, err := c.convertToConfig(obj)
+		if err != nil {
+			continue
+		}
+
+		var oldEpVersion string
+		c.configStoreMu.Lock()
+		namedConf, ok := c.configStore[conf.Namespace]
+		if ok {
+
+			if namedConf[conf.Name] != nil {
+				oldEpVersion = version(namedConf[conf.Name].Annotations, endpointKey)
+			}
+			namedConf[conf.Name] = conf
+		} else {
+			c.configStore[conf.Namespace] = map[string]*model.Config{
+				conf.Name: conf,
+			}
+		}
+		c.configStoreMu.Unlock()
+
+		if svcChanged {
+			continue
+		}
+
+		newEpVersion := version(conf.Annotations, endpointKey)
+		if oldEpVersion != newEpVersion {
+			if err := c.edsUpdate(conf); err != nil {
+				log.Warnf("edsUpdate: %v", err)
+			}
+		}
+	}
+	if svcChanged {
+		c.options.XDSUpdater.ConfigUpdate(true)
+	}
+}
+
+func (c *IncrementalController) edsUpdate(config *model.Config) error {
+	se, ok := config.Spec.(*networking.ServiceEntry)
+	if !ok {
+		return errors.New("edsUpdate: wrong type")
+	}
+	istioEndpoints := convertEndpoints(se, config.Name, config.Namespace)
+	hostname := hostName(config.Name, config.Namespace, c.options.DomainSuffix)
+	return c.options.XDSUpdater.EDSUpdate("MCP", hostname, istioEndpoints)
+}
+
+func (c *IncrementalController) configExist(metadataName string) *model.Config {
+	namespace, name := extractNameNamespace(metadataName)
+	c.configStoreMu.Lock()
+	defer c.configStoreMu.Unlock()
+	configs, ok := c.configStore[namespace]
+	if !ok {
+		return nil
+	}
+	conf, ok := configs[name]
+	if !ok {
+		return nil
+	}
+	return conf
+}
+
+func (c *IncrementalController) serviceChange(resources []*sink.Object) bool {
+	for _, obj := range resources {
+		conf := c.configExist(obj.Metadata.Name)
+		if conf == nil {
+			// if config does not exist send it down configUpdate path
+			return true
+		}
+		oldVersion := version(conf.Annotations, serviceKey)
+		newVersion := version(obj.Metadata.Annotations, serviceKey)
+		if oldVersion != newVersion {
+			return true
+		}
+	}
+	return false
+}
+
+func convertEndpoints(se *networking.ServiceEntry, cfgName, ns string) (endpoints []*model.IstioEndpoint) {
+	for _, ep := range se.Endpoints {
+		for portName, port := range ep.Ports {
+			ep := &model.IstioEndpoint{
+				Address:         ep.Address,
+				EndpointPort:    port,
+				ServicePortName: portName,
+				Labels:          ep.Labels,
+				UID:             cfgName + "." + ns,
+				Network:         ep.Network,
+				Locality:        ep.Locality,
+				LbWeight:        ep.Weight,
+				// ServiceAccount:??
```

Comment:
@costinm @andraxylia now that service accounts have been pulled up to the top-level of `ServiceEntry`, what should @Nino-K do here?